src/nvim/screen.c

```diff
@@ -2231,34 +2151,11 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
 
     has_decor = decor_redraw_line(buf, lnum-1, &decor_state);
 
-    for (size_t k = 0; k < kv_size(*providers); k++) {
-      DecorProvider *p = kv_A(*providers, k);
-      if (p && p->redraw_line != LUA_NOREF) {
-        FIXED_TEMP_ARRAY(args, 3);
-        args.items[0] = WINDOW_OBJ(wp->handle);
-        args.items[1] = BUFFER_OBJ(buf->handle);
-        args.items[2] = INTEGER_OBJ(lnum-1);
-        if (provider_invoke(p->ns_id, "line", p->redraw_line, args, true)) {
-          has_decor = true;
-        } else {
-          // return 'false' or error: skip rest of this window
-          kv_A(*providers, k) = NULL;
-        }
-
-        win_check_ns_hl(wp);
-      }
-    }
+    providers_invoke_line(wp, providers, lnum-1, &has_decor, &provider_err);
 
     if (provider_err) {
-      Decoration err_decor = DECORATION_INIT;
-      int hl_err = syn_check_group(S_LEN("ErrorMsg"));
-      kv_push(err_decor.virt_text,
-              ((VirtTextChunk){ .text = provider_err,
-                                .hl_id = hl_err }));
-      err_decor.virt_text_width = mb_string2cells((char_u *)provider_err);
-      decor_add_ephemeral(lnum-1, 0, lnum-1, 0, &err_decor);
+      provider_err_virt_text(lnum, provider_err);
       provider_err = NULL;
-      has_decor = true;
```

Comment:
I think this is still necessary.

Reply:
I've incorporated this into `decor_provider_invoke_line`